src/github.com/matrix-org/dendrite/roomserver/input/events.go

```diff
@@ -9,18 +9,33 @@ import (
 // A RoomEventDatabase has the storage APIs needed to store a room event.
 type RoomEventDatabase interface {
 	// Stores a matrix room event in the database
-	StoreEvent(event gomatrixserverlib.Event, authEventNIDs []int64) error
+	StoreEvent(event gomatrixserverlib.Event, authEventNIDs []types.EventNID) (types.RoomNID, types.StateAtEvent, error)
 	// Lookup the state entries for a list of string event IDs
 	// Returns a sorted list of state entries.
-	// Returns a error if the there is an error talking to the database
+	// Returns an error if the there is an error talking to the database
 	// or if the event IDs aren't in the database.
 	StateEntriesForEventIDs(eventIDs []string) ([]types.StateEntry, error)
 	// Lookup the numeric IDs for a list of string event state keys.
 	// Returns a map from string state key to numeric ID for the state key.
-	EventStateKeyNIDs(eventStateKeys []string) (map[string]int64, error)
+	EventStateKeyNIDs(eventStateKeys []string) (map[string]types.EventStateKeyNID, error)
 	// Lookup the Events for a list of numeric event IDs.
 	// Returns a sorted list of events.
-	Events(eventNIDs []int64) ([]types.Event, error)
+	Events(eventNIDs []types.EventNID) ([]types.Event, error)
+	// Lookup the state of a room at each event for a list of string event IDs.
+	// Returns a sorted list of state at each event.
+	// Returns an error if there is an error talking to the database
+	// or if the room state for the event IDs aren't in the database
+	StateAtEventIDs(eventIDs []string) ([]types.StateAtEvent, error)
+	// Lookup the numeric state data IDs for the each numeric state ID
+	// The returned slice is sorted by numeric state ID.
+	StateDataNIDs(stateNIDs []types.StateSnapshotNID) ([]types.StateDataNIDList, error)
```

Comment:
Also saying "sorted by numeric state ID" is vague: Sorted by the order they are in the array, or sorted by the actual value of each NID?